source/manual/intro-to-govuk-docker.html.md

```diff
@@ -0,0 +1,160 @@
+---
+title: Intro to Docker and govuk-docker
+section: Docker
+layout: manual_layout
+parent: "/manual.html"
+last_reviewed_on: 2019-07-01
+review_in: 6 months
+owner_slack: "#govuk-developers"
+---
+
+## Intro
+If you are new to Docker here's a quick intro in the context of the GOV.UK stack. There is also a wrapper for a lot of the docker commands in govuk-docker, which we will look at later on.
+
+
+### Install Docker
+
+*   Download Docker [https://www.docker.com/get-started](https://www.docker.com/get-started)
+*   Docker for Mac intro [https://docs.docker.com/docker-for-mac/](https://docs.docker.com/docker-for-mac/)
+
+### Video explaining the concepts in Docker
+
+[https://www.youtube.com/watch?v=i7yoXqlg48M](https://www.youtube.com/watch?v=i7yoXqlg48M) - first 30 minutes
+
+## Learning docker with content-publisher
+We'll run a few things here to get the [content-publisher] up and running in Docker. This is a convoluted example but it will help to explain the concepts involved.
+
+### Run a container from an image
+
+*   Make sure you have docker running, then run from the root of the [content-publisher] project:
+
+```shell
+docker run -it --rm ubuntu
+```
+
+#### What are Docker images and a container?
+The `ubuntu` is the "image" that we specify. Using object oriented programming as an analogy, an image is like a "class". When we "instantiate" an instance of a class, these are called "containers".
+
+Images are downloaded from the Docker Registry. The containers are run as a process on our host.
+
+Another image is ruby:
+
+```shell
+docker run -it --rm ruby:2.6.3 bash
+```
+
+Here we get a ruby image and specifying which version we want. The `bash` at the end means we get a bash shell.
+
+### Volumes
+
+* Add a volume by running:
+
+```shell
+docker run -it --rm -v $(pwd):/app ruby:2.6.3 bash
+```
+
+This will map your current directory (being the root of the [content-publisher] project) to `/app` inside of the container.
+
+If you were to run bundle install for the app, the next time we spin up a container all of the gems would need re-installing.
+
+#### Persistent volumes
+
+Docker allows you create separate volumes for persistent data.
+
+* Create a persistent volume for our gems:
+
+```shell
+docker volume create content-publisher-bundle
+```
+
+*   Run docker with that volume:
+
+```shell
+docker run -it --rm -v $(pwd):/app -v content-publisher-bundle:/usr/local/bundle ruby:2.6.3 bash
+```
+
+*   Then installing gems:
+```shell
+cd /app
+bundle install
```

Comment:
I guess it might confuse, but it could be useful to illustrate what ‘persisted’ means by suggesting people exit and re-enter a container.

Reply:
I think people would get the idea of persistent data, but if not the person who doesn't can raise the PR. As you say this may be confusing so I'm keeping it simple for now.